mantidproject/mantid
# Fix unintentional modification of container workspace and make rebinning optional in ApplyPaalmanPingsCorrection

[`ApplyPaalmanPingsCorrection`](http://docs.mantidproject.org/nightly/algorithms/ApplyPaalmanPingsCorrection-v1.html) was accidentally applying [`RebinToWorkspace`](http://docs.mantidproject.org/nightly/algorithms/RebinToWorkspace-v1.html) to the input *CanWorkspace*. This PR fixes the behavior and also makes the rebinning optional as it is not always desirable. For example, if both *SampleWorkspace* and *CanWorkspace* have varying bins, `RebinToWorkspace` would rebin the container workspace incorrectly.

**To test:**

Before applying these changes, running
```python
import numpy

xs = numpy.array([0.0, 1.0, 0.0, 1.1])
ys = numpy.array([2.2, 3.3])
sample_1 = CreateWorkspace(DataX=xs, DataY=ys, NSpec=2, UnitX='Wavelength')
ys = numpy.array([0.11, 0.22])
container_1 = CreateWorkspace(DataX=xs, DataY=ys, NSpec=2, UnitX='Wavelength')
corrected = ApplyPaalmanPingsCorrection(SampleWorkspace=sample_1, CanWorkspace=container_1)
print(container_1)
```
would cause `container_1` to get rebinned, and sometimes raise an exception at the last `print` statement. This should not happen anymore.

Further, to test the optional rebinning, one could run the following script:
```python
import numpy

xs = numpy.array([0.0, 1.0, 0.0, 1.1])
ys = numpy.array([2.2, 3.3])
sample_1 = CreateWorkspace(DataX=xs, DataY=ys, NSpec=2, UnitX='Wavelength')
xs = numpy.array([-0.1, 0.5, 1.1, -0.1, 0.5, 1.2])
ys = numpy.array([0.11, 0.22, 0.33, 0.44])
container_1 = CreateWorkspace(DataX=xs, DataY=ys, NSpec=2, UnitX='Wavelength')
corrected = ApplyPaalmanPingsCorrection(SampleWorkspace=sample_1, CanWorkspace=container_1, RebinCanToSample='Rebin Can')
```
In the above, setting *RebinCanToSampe* to 'Do Not Rebin Can' should make the algorithm fail due to incompatible sample and container workspaces.

Fixes #18383.

*The release notes were updated accordingly.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####


##### Functional Tests #####


- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
